xoinvader/collision.py

```diff
@@ -1,49 +1,184 @@
-    print(settings_manager._smap)
-    s1.notifySettingsManager()
-    print(settings_manager._smap)
+"""Collision detection system and component module."""
+
+import weakref
+
+from xoinvader.utils import Point
+
+
+class TypePair(object):
+    """Class for hashable unordered string pairs.
+
+    Used as collision dictionary keys, containing pair of collider types. It's
+    set-like, i.e. TypePair(a, b) == TypePair(b, a) and their hashes are equal
+    too.
+
+    :param str a: first collider type
+    :param str b: second collider type
+    """
+
+    def __init__(self, a, b):
+        a, b = sorted([a, b])
+        self._a = a
+        self._b = b
+        self._ab = a + '_' + b
+
+    @property
+    def a(self):
+        return self._a
+
+    @property
+    def b(self):
+        return self._b
+
+    def __eq__(self, other):
+        return self._ab == other._ab
+
+    def __hash__(self):
+        return hash(self._ab)
+
+
+class CollisionManager(object):
+    """Class for collision detection between known components.
+
+    To process collisions, first update the positions of all objects of
+    interest, then call `update` method. It will traverse all registered
+    collisions (between pairs of types) and call appropriate handlers in order
+    they were registered for the two types of colliding objects.
+
+    If you just want to check, if two Colliders collide, call `check_collision`
+    on them.
+    """
+
+    def __init__(self):
+        if Collider.__manager__ is None:
+            def null_manager(_):
+                Collider.__manager__ = None
+            Collider.__manager__ = weakref.ref(self, null_manager)
+        self._colliders = []
+        self._collisions = {}
+
+    def _add(self, collider):
+        self._colliders.append(collider)
+
+    def update(self):
+        """Detect and process all collisions."""
+        for pair in self._collisions:
+            c_a = [n for n in self._colliders if n.type == pair.a]
+            c_b = [n for n in self._colliders if n.type == pair.b]
+            for cldr_a in c_a:
+                for cldr_b in c_b:
+                    maybe_coll_rect = self.check_collision(cldr_a, cldr_b)
+                    if maybe_coll_rect:
+                        callbacks = self._collisions[pair]
+                        for cb in callbacks:
+                            if cb:
+                                cb(maybe_coll_rect)
+
+    def check_collision(self, col1, col2):
+        """Check collisions between two colliders.
+
+        Returns `None` if no collision occured, or returns rectangle of
+        overlapping region between collider maps.
+
+        :param col1: first collider
+        :type col1: :class:`Collider`
+        :param col2: second collider
+        :type col2: :class:`Collider`
+        :rtype: tuple of two :class:`Point`
+        """
+
+        w1 = max(map(len, col1.map))
+        h1 = len(col1.map)
+        tl1 = col1.pos
+        rb1 = tl1 + Point(w1, h1)
```

Comment:
I guess it will do. Actually I after all might like it even over my improved version.